### PR TITLE
AVRO-3256: IDL type reference with annotation throws error

### PR DESCRIPTION
**Pull request by opwvhk**
Previous versions would alter the referenced type when encountering an
annotation on (for example) a field type. This change makes references
read-only.

Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-3256
  - ~In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).~

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
  - {{org.apache.avro.compiler.idl.TestReferenceAnnotationNotAllowed}}
  - Updates to {{simple.avdl}} and {{simple.avpr}} (as described in the Jira issue)
  - Updates to other test schemata that contained the same bug

### Commits

- [X] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj [hunk 1]
```diff
@@ -1141,7 +1141,7 @@ List<String> EnumBody():
 }
 {
   "{"
-  [ EnumConstant(symbols) ( LOOKAHEAD(2) "," EnumConstant(symbols) )* ]
```
Comment (opwvhk) [inline lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj:1144]: This lookahead was never needed.

File: lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj [hunk 2]
```diff
@@ -1317,17 +1317,9 @@ void FieldDeclaration(List<Field> fields):
...
-  // or just on field declarations as done here
-
-  ( SchemaProperty(props) )*
   type = Type()
   VariableDeclarator(type, fields) ( "," VariableDeclarator(type, fields) )*
   ";"
-  {
-    for (String key : props.keySet())
-      Accessor.addProp(type, key, props.get(key));
-  }
```
Comment (opwvhk) [inline lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj:1330]: The Type() production already handles properties; no need to add them here as well.

File: lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj [hunk 3]
```diff
@@ -1440,16 +1432,27 @@ Schema Type():
 {
 
     ( SchemaProperty(props) )*
+  s = UnannotatedType(props)
```
Comment (opwvhk) [inline lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj:1435]: Added to use it for message return types (the grammar attaches annotations before the return type on the message).

Using this production here makes sure the other lookahead (in line 1444 -> 1449) is no longer needed.

File: lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj [hunk 4]
```diff
@@ -1440,16 +1432,27 @@ Schema Type():
...
+  }
+}
+
+Schema UnannotatedType(Map<String, JsonNode> props):
+{
+  Schema s;
+}
+{
   (
-      LOOKAHEAD(2) s = ReferenceType()
```
Comment (opwvhk) [inline lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj:1444]: This lookahead is no longer needed, due to UnannotatedType(...).

File: lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj [hunk 5]
```diff
@@ -1543,9 +1546,8 @@ Schema ResultType():
   Schema schema;
 }
 {
-  LOOKAHEAD(2)
     "void"          { return Schema.create(Type.NULL); }
-  | schema = Type() { return schema; }
```
Comment (opwvhk) [inline lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj:1548]: This is the "magic" bit that ensures the grammer needs no lookahead anymore.

**Review by RyanSkraba — APPROVED** [review]
Hey, thanks so much for this contribution!  I've been reading up on JavaCC so I can help review some of these important PRs -- your attention to this part of the code base is really appreciated!  I especially appreciate your notes on the PR to help explain what's going on.

File: doc/src/content/xdocs/idl.xml [hunk 6]
```diff
@@ -429,50 +429,62 @@ record MyRecord {
...
+* Header with license information.
+*/
+
 /**
  * An example protocol in Avro IDL
  */
 @namespace("org.apache.avro.test")
 protocol Simple {
-
+  /** Documentation for the enum type Kind */
```
[the remark this answers is not in the document]
Comment (opwvhk) [reply doc/src/content/xdocs/idl.xml:449]: The parser supports doc comments, but only trims them; any * prefixes on lines are kept.
I intend to improve upon this in a PR that'll also ignore doc comments halfway declarations. But I need PR #1377 for this, as I want the tools to emit warnings when this happens.

**Pull request merged**